File: leetcode/Univalued_Binary_Tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def _init__(self):
        self.val = None

    def isUnivalTree(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        if self.val is None:
            self.val = root.val
        else:
            if root.val != self.val:
                return False
        if root.left and root.right:
            return self.isUnivalTree(root.left) and self.isUnivalTree(root.right)
        elif root.left:
            return self.isUnivalTree(root.left)
        elif root.right:
            return self.isUnivalTree(root.right)
        return True

File: leetcode/univaluedBinaryTree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */


var isUnivalTree = function(root) {
    const matchVal = root.val;
    const traverse = (root, val) => {
        // console.log(`root.val: ${root.val}, val: ${val}`);
        if (root.val !== val) return false;
        let res = true;
        if (root.left) {
            if (!traverse(root.left, val)) return false;
        }
        if (root.right) {
            if (!traverse(root.right, val)) return false;
        }
        return true;
    };
    return traverse(root, matchVal);
};
